### PR TITLE
ACS-3183 POST rules, GET, PUT and DELETE rule API definition.

### PR DESCRIPTION
**Pull request by tpage-alfresco**
Initial spec for Rules CRUD. More PRs to follow, eg. to add link/unlink rulesets etc.

Also some small changes to the GET rules API, the largest being the introduction
of a new type for action definitions (one that doesn't include a target node id).

### CODE REVIEW

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 1]
```diff
@@ -8472,7 +8472,48 @@ paths:
...
+    post:
+      x-alfresco-since: "7.3.0"
+      tags:
+        - rules
+      summary: Create a rule
+      description: |
+        **Note:** this endpoint is available in Alfresco 7.3 and newer versions.
 
+        Creates a rule on folder **nodeId**.
+      operationId: createComment
```
Comment (tpage-alfresco) [inline src/main/webapp/definitions/alfresco-core.yaml:8484]: This still mentions comments

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 2]
```diff
@@ -10758,12 +10809,68 @@ definitions:
...
+
+          If the field is omitted then the rule will apply to all nodes.
         type: '#/definitions/CompositeConditionDefinition'
       actions:
         description: The actions for the rule
         type: array
         items:
-          $ref: '#/definitions/ActionDefinition'
+          $ref: '#/definitions/ActionBodyExecTemplate'
+  RuleBody:
```
Comment (krdabrowski) [inline src/main/webapp/definitions/alfresco-core.yaml:10820]: The `RuleBody` is almost the same as `Rule`, isn't it possible to use one object in both cases? Or introduce inheritance?

Comment (tpage-alfresco) [reply src/main/webapp/definitions/alfresco-core.yaml:10820]: This wasn't possible last time I checked.  There's syntax for `allOf`, `anyOf` and `oneOf` which allow this sort of reuse, but IIRC there was some reason we couldn't use it.  @montgolfiere Do you remember any more about this issue?

**Comment by tpage-alfresco** [comment]
Added API definitions for GET/PUT/DELETE a single rule and updated everything to include the concept of a ruleset (so now we have the three level path `/nodes/{nodeId}/rulesets/{rulesetId}/rules/{ruleId}`).

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 3]
```diff
@@ -8435,7 +8447,7 @@ paths:
         '404':
           description: |
             **actionDefinitionId** or **targetId** does not exist
-  '/nodes/{nodeId}/rules':
+  '/nodes/{nodeId}/rulesets/{rulesetId}/rules':
```
Comment (montgolfiere) [inline src/main/webapp/definitions/alfresco-core.yaml:8450]: If this is -default- (or similar) then should be stated in each method description (alongside folder "nodeId").

**Review by montgolfiere — COMMENTED** [review]
Thanks @tpage-alfresco. It looks like this coming together nicely. I've a few minor initial comments

We also will need to add support for:

- re-order rules
- linked rulesets (link , unlink, view - who I link to, who links to me)
- disable / enable rule inheritance

I realise we may decide to add incrementally via separate PRs (eg. see https://github.com/Alfresco/rest-api-explorer/pull/159 ).

Also consider new features / stretch goals such as:

- OR conditions (as per composite nested grouping)
- is null / is not null
- copy rule (presumably to another folder)

**Review by montgolfiere — APPROVED** [review]
As discussed, I think we can merge this initial Rules CRUD spec proposal on the basis that:

- this is still experimental ( & hence subject to change ) targeting currently planned ACS Content Repo 7.3.0
- we will have follow-on PRs for other required APIs (eg. link / unlink rulesets as per Draft https://github.com/Alfresco/rest-api-explorer/pull/159) and any other identified gaps &/or new features (see also https://github.com/Alfresco/rest-api-explorer/pull/161#pullrequestreview-1027420129 )
- we will keep up-to-date as we build out the API impl (eg. [first step](https://github.com/Alfresco/alfresco-community-repo/pull/1193)) as well as end-to-end API test scenarios and client usage (eg. ACA/ADW UIs)
- we will review with ACS architects (once we have the ability to demonstrate most/all end-to-end scenarios)

**Review by krdabrowski — APPROVED** [review]

**Pull request merged**